src/html/Quoting.cc

```diff
@@ -9,81 +9,74 @@
 #include "squid.h"
 #include "html/Quoting.h"
 
-/*
- *  HTML defines these characters as special entities that should be quoted.
- */
-static struct {
-    unsigned char code;
-    const char *quote;
-} htmlstandardentities[] =
+#include <array>
+#include <cstring>
+
+class HtmlSpecialCharacters {
+private:
+    std::array<const char *, 256> escapeMap = {};
 
+public:
+    HtmlSpecialCharacters();
+    ~HtmlSpecialCharacters();
+    const char *operator[](unsigned char ch) const { return escapeMap[ch]; }
+};
+HtmlSpecialCharacters::HtmlSpecialCharacters()
 {
-    /* NOTE: The quoted form MUST not be larger than 6 character.
-     * see close to the MemPool commend below
+    static std::array<std::pair<unsigned char, const char *>, 5> const escapePairs = {
+        std::make_pair('<', "&lt;"),
+        std::make_pair('>', "&gt;"),
+        std::make_pair('"', "&quot;"),
+        std::make_pair('&', "&amp;"),
+        std::make_pair('\'', "&apos;")
+    };
+    const size_t maxEscapeLength = 7;
+    /* Encode control chars just to be on the safe side, and make
+     * sure all 8-bit characters are encoded to protect from buggy
+     * clients
      */
-    {
-        '<', "&lt;"
-    },
-    {
-        '>', "&gt;"
-    },
-    {
-        '"', "&quot;"
-    },
-    {
-        '&', "&amp;"
-    },
-    {
-        '\'', "&#39;"
-    },
-    {
-        0, NULL
+    for (uint32_t ch = 1; ch < 256; ++ch) {
```

Comment:
This misses out the `&#0;` character reference for ASCII `NUL`.

Reply:
Yes. However, as the input to the function is null-terminated strings, we have no way to receive that input

Reply:
You are both right. Please include NUL because

* HtmlSpecialCharacters class is not documented to be specific/internal to html_quote(const char *);
* Including it will reduce the probability of bugs in future refactoring changes where input becomes something other than a c-string.

```suggestion
    for (uint32_t ch = 0; ch < 256; ++ch) {
```

Reply:
done, resolving